src/algorithms/tonal/pitchyinprobabilities.cpp

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright (C) 2006-2018  Music Technology Group - Universitat Pompeu Fabra
+ *
+ * This file is part of Essentia
+ *
+ * Essentia is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU Affero General Public License as published by the Free
+ * Software Foundation (FSF), either version 3 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the Affero GNU General Public License
+ * version 3 along with this program.  If not, see http://www.gnu.org/licenses/
+ */
+
+#include "pitchyinprobabilities.h"
+#include "essentiamath.h"
+
+using namespace std;
+using namespace essentia;
+using namespace standard;
+
+
+const char* PitchYinProbabilities::name = "PitchYinProbabilities";
+const char* PitchYinProbabilities::category = "Pitch";
+const char* PitchYinProbabilities::description = DOC("This algorithm estimates the fundamental frequencies, their probabilities given the frame of a monophonic music signal. It is a part of the implementation of the probabilistic Yin algorithm [1].\n"
+"\n"
+"An exception is thrown if an empty signal is provided.\n"
+"\n"
+"References:\n"
+"  [1] M. Mauch and S. Dixon, \"pYIN: A Fundamental Frequency Estimator\n"
+"  Using Probabilistic Threshold Distributions,\" in Proceedings of the\n"
+"  IEEE International Conference on Acoustics, Speech, and Signal Processing\n"
+"  (ICASSP 2014)Project Report, 2004");
+
+
+void PitchYinProbabilities::configure() {
+  _frameSize = parameter("frameSize").toInt();
+  _sampleRate = parameter("sampleRate").toReal();
+  _lowAmp = parameter("lowAmp").toReal();
+  _precisetime = parameter("precisetime").toBool();
+
+  _yin.resize(_frameSize/2 + 1);
+  _peakProb.resize(_yin.size());
+
+  // Pre-processing
+  _FFT->configure("negativeFrequencies", true,
+                  "size", _frameSize);
+  _IFFT->configure("size", _frameSize);
+}
+
+Real PitchYinProbabilities::parabolicInterpolation(const std::vector<Real> yinBuffer, const size_t tau, const size_t yinBufferSize) {
+  // this is taken almost literally from Joren Six's Java implementation
+  if (tau == yinBufferSize) // not valid anyway.
+  {
+    return static_cast<Real>(tau);
+  }
+  
+  Real betterTau = 0.0;
+  if (tau > 0 && tau < yinBufferSize-1) {
+    Real s0, s1, s2;
+    s0 = yinBuffer[tau-1];
+    s1 = yinBuffer[tau];
+    s2 = yinBuffer[tau+1];
+      
+    Real adjustment = (s2 - s0) / (2 * (2 * s1 - s2 - s0));
+      
+    if (abs(adjustment)>1) adjustment = 0;
+      
+    betterTau = tau + adjustment;
+  } else {
+    betterTau = tau;
+  }
+  return betterTau;
+}
+
+Real PitchYinProbabilities::sumSquare(const std::vector<Real> signal, const size_t start, const size_t end) {
```

Comment:
This function is useful for more applications, so it should be placed into essentiamath.h

Reply:
This looks like a generic template function that more algorithms can take advantage of if it is placed to ``essentiamath.h``. This header contains many functions operating on vectors (for example, [sum](https://github.com/MTG/essentia/blob/master/src/essentia/essentiamath.h#L103))